File: runtime/go/value_adapter.go
package runtime

import (
	"context"
	"encoding/json"
)

// ValueAdapter pairs a declaration with its two primitive conversions. A
// reusable adapter retains no operation context or lifetime. When NeedsContext
// is true, conversion runs inside the supplied ValueEnvironment's matching
// Export, Import or Publish boundary, using the active context it provides.
type ValueAdapter[T any] struct {
	Binding      TypeBinding
	NeedsContext bool
	Export       func(context.Context, T) (json.RawMessage, error)
	Import       func(context.Context, json.RawMessage) (T, error)
}

// ValueEnvironment supplies operation-local conversion effects. It belongs to
// the consumer's chosen component; the runtime never interprets its context.
// Select chooses an outgoing context, and Child derives an incoming lifetime.
// Build callbacks are synchronous and must pass their active context through
// every nested conversion. Publish completes the build before attempting send.
type ValueEnvironment interface {
	Select(context.Context) (context.Context, error)
	Child(context.Context) (context.Context, error)
	Export(context.Context, func(context.Context) (json.RawMessage, error)) (json.RawMessage, error)
	Import(context.Context, func(context.Context) error) error
	Publish(context.Context, func(context.Context) (json.RawMessage, error), func(json.RawMessage) (json.RawMessage, error)) (json.RawMessage, error)
}

// JSONAdapter validates ordinary data in both directions. It neither consults
// the context nor requires a conversion environment.
func JSONAdapter[T any]() ValueAdapter[T] {
	binding := TypeArgument[T]()
	return ValueAdapter[T]{
		Binding: binding,
		Export: func(_ context.Context, value T) (json.RawMessage, error) {
			raw, err := MarshalJSON(value)
			if err == nil {
				err = binding.Schema.ValidateExpressionRaw(binding.Type, raw)
			}
			return raw, err
		},
		Import: func(_ context.Context, raw json.RawMessage) (T, error) {
			var value T
			if err := binding.Schema.ValidateExpressionRaw(binding.Type, raw); err != nil {
				return value, err
			}
			err := json.Unmarshal(raw, &value)
			return value, err
		},
	}
}

File: runtime/ts/src/value-adapter.ts
import type { TypeBinding } from './validate.ts';

/** Primitive conversion with no retained operation context or lifetime.
 * Acquiring conversions run inside the chosen environment's matching boundary,
 * using its active build context throughout the whole value traversal.
 */
export interface ValueAdapter<T> {
  readonly binding: TypeBinding;
  readonly needsContext: boolean;
  readonly export: (context: unknown, value: T) => unknown;
  readonly import: (context: unknown, value: unknown) => T;
}

/** Consumer-supplied effects; the runtime never interprets their contexts. */
export interface ValueEnvironment {
  select(context: unknown): unknown;
  child(context: unknown): unknown;
  export(context: unknown, build: (view: unknown) => unknown): unknown;
  import<T>(context: unknown, build: (view: unknown) => T): T;
  publish<T>(context: unknown, build: (view: unknown) => unknown, publish: (raw: unknown) => Promise<T>): Promise<T>;
}

type CarriesContext<T, Seen = never> = [T] extends [Seen]
  ? false
  : T extends (...args: never[]) => unknown
    ? true
    : T extends readonly (infer Item)[]
      ? CarriesContext<Item, Seen | T>
      : T extends object
        ? true extends { [K in keyof T]: CarriesContext<T[K], Seen | T> }[keyof T]
          ? true
          : false
        : false;

/** Specialization preserves an operation's context requirement without choosing its provider. */
export type ValueContext<T, Context> = true extends CarriesContext<T> ? Context & { valueContext: unknown } : Context;
export type ValueOptions<T, Options> = true extends CarriesContext<T> ? Options & { valueContext?: unknown } : Options;

/** Data validates in both directions without consulting any conversion context. */
export function jsonAdapter<T>(binding: TypeBinding): ValueAdapter<T> {
  return {
    binding,
    needsContext: false,
    export(_context, value) {
      binding.validate(binding.type, value, '$', binding.slots);
      return value;
    },
    import(_context, value) {
      binding.validate(binding.type, value, '$', binding.slots);
      return value as T;
    },
  };
}
